sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/CentralMomentAgg.scala

```diff
@@ -57,173 +55,170 @@ abstract class CentralMomentAgg(child: Expression) extends ImperativeAggregate w
 
   override def dataType: DataType = DoubleType
 
-  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType)
+  override def inputTypes: Seq[AbstractDataType] = Seq(DoubleType)
+
+  protected val count = AttributeReference("count", DoubleType, nullable = false)()
+  protected val avg = AttributeReference("avg", DoubleType, nullable = false)()
+  protected val m2 = AttributeReference("m2", DoubleType, nullable = false)()
+  protected val m3 = AttributeReference("m3", DoubleType, nullable = false)()
+  protected val m4 = AttributeReference("m4", DoubleType, nullable = false)()
+
+  override val aggBufferAttributes = Seq(count, avg, m2, m3, m4).take(momentOrder + 1)
+
+  override val initialValues: Seq[Expression] = Seq(
+    /* count = */ Literal(0.0),
+    /* avg = */ Literal(0.0),
+    /* m2 = */ Literal(0.0),
+    /* m3 = */ Literal(0.0),
+    /* m4 = */ Literal(0.0)
+  ).take(momentOrder + 1)
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    val n = count + Literal(1.0)
+    val delta = child - avg
+    val deltaN = delta / n
+    val newAvg = avg + deltaN
+    val newM2 = m2 + delta * (delta - deltaN)
+
+    val delta2 = delta * delta
+    val deltaN2 = deltaN * deltaN
+    val newM3 = m3 - Literal(3.0) * deltaN * newM2 + delta * (delta2 - deltaN2)
+
+    val newM4 = m4 - Literal(4.0) * deltaN * newM3 - Literal(6.0) * deltaN2 * newM2 +
+      delta * (delta * delta2 - deltaN * deltaN2)
+
+    if (child.nullable) {
```

Comment:
It might be easier to understand if we write

~~~scala
trimHigherOrder(
  If(IsNull(child),
    Seq(count, avg, m2, m3, m4),
    Seq(n, newAvg, newM2, newM3, newM4)))
~~~

Please let me know if it doesn't work. Btw, I guess `if (child.nullable)` is automatically checked by `IsNull` (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala#L74). Then we can remove that check to simplify the code here.